### ISSUE TITLE
twisted.web2.channel.scgi erroneously sends bare LF

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4032]
|<img alt="TorryCn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| TorryCn reported|
|-|-|
|Trac ID|trac#4032|
|Type|defect|
|Created|2009-09-23 04:28:42Z|
|Branch|https://github.com/twisted/twisted/tree/scgi-crlf-4032|

When  twisted.web.channel.scgi work with lighttpd, 
There is a line be insert before the webpage content.

The problem at http://twistedmatrix.com/trac/browser/tags/releases/twisted-8.2.0/twisted/web2/channel/scgi.py#L18

l.append("Status: %s %s\n" % (code, code_message))

change it to

l.append("Status: %s %s\r\n" % (code, code_message))

insert '\r' before '\n'

The problem is clean.

My English is poor, if you can't understand what I say. please contact me. torrycn@...

Attachments:

* [my-scgi-patch.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/9f2/9f26a7e30c11faa76d3172144c94f8d0bb891cf6/1017058920c789225cf0082f1d9e8abf06be22cf.diff) (587 bytes) - added by TorryCn on 2009-10-14 08:51:13Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__4032 4032
type__defect defect
reporter__TorryCn TorryCn
priority__normal normal
milestone__ 
branch__branches_scgi_crlf_4032 branches/scgi-crlf-4032
branch_author__jml jml
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1253680122000000 1253680122000000
changetime__1261288820000000 1261288820000000
version__None None
owner__ 
cc__therve cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="ivank's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|ivank set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="TorryCn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|TorryCn set owner to TorryCn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to TorryCn|
|-|-|

Thanks a lot for your patch, and sorry for not answering you sooner. This patch looks correct, but it would need a test to be checked in. Do you think you could write such a test? It seems scgi has some coverage, it shouldn't be too tricky to write one. Thanks!

**Comment by twisted-trac** [comment]
|<img alt="TorryCn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|TorryCn commented|
|-|-|

I am a beginner. I want do something for the problem.  Can you guide me please?

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

(In [27783]) Branching to 'scgi-crlf-4032'

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_7"></a>|@jml commented|
|-|-|

(In [27784]) Add TorryCn's patch, along with a test. refs [#4032](https://github.com/twisted/twisted/issues/4032)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_9"></a>|@jesstess set owner to @jml|
|-|-|

The patch and test look fine. On the docstrings:
 * "handle a L{SCGIChannelRequest.writeHeaders}" ==> "handle an L{SCGIChannelRequest.writeHeaders}".
 * Digging through RFC 1945, it references RFC 822 on header formatting and section 3.1 of 822 seems to imply that it's really CR and/or LF instead of CRLF. It may be worth mentioning in the test docstring that we're adding the carriage return because it's the standard in practice even if not by the letter of the RFC (#twisted seemed unconcerned about this, though).

Other than that, looks good, please apply!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml set status to `closed`|
|-|-|

(In [27837]) Use consistent separators in headers in SCGI output from web2.

 * Authors: TorryCn, jml
 * Reviewer: jesstess
 * Fixes [#4032](https://github.com/twisted/twisted/issues/4032)

Use CR LF always, rather than just CR after the status and CR LF elsewhere.
This fixes a display problem when using Twisted SCGI with lighttpd.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
